lib/Mojolicious/Guides/Testing.pod

```diff
@@ -0,0 +1,814 @@
+L<Mojo::UserAgent::Transactor>) to marshal hash and array references into JSON
+documents, and then send them as a WebSocket message. We wait (block) for a
+response from the server with C<message_ok>. Then because we're expecting a
+JSON document back, we can leverage C<json_message_ok> which parses the
+WebSocket response body and returns an object we can access through
+L<Mojo::JSON::Pointer> syntax. Then we close (and test) our WebSocket
+connection.
+
+Testing WebSocket servers does not get any simpler than with L<Test::Mojo>.
+
+=head2 Extending L<Test::Mojo>
+
+In some situations you may find your tests have a lot of boilerplate, making
+tests cluttered and hard to follow. For example let's say we have an RPC-based
+web application where the HTTP method is always C<POST> and the remote
+procedure name is in the URL, and we use HTTP Basic authentication with
+username C<me> and password C<secretly>.
+
+All of our requests follow this basic pattern:
+
+  POST /<service>/<command>
+  Authorization: Basic xxxxx
+
+  {"json":"payload"}
+
+For example:
+
+  POST /bedroom-light/power-on
+  Authorization: Basic bWU6c2VjcmV0bHkK=
+
+  {"level":"50%"}
+
+Or:
+
+  POST /blender/spin-blades
+  Authorization: Basic bWU6c2VjcmV0bHkK=
+
+  {"speed":"whip"}
+
+Or:
+
+  POST /thermostat/heat
+  Authorization: Basic bWU6c2VjcmV0bHkK=
+
+  {"temperature":"23°C"}
+
+To test this kind of web application, we'd need to do something like this for
+every test:
+
+  $t->post_ok('/thermostat/heat',
+              {Authorization => 'Basic ' . b64_encode 'me:secretly', ''},
+              json => {temperature => '23°C'})
+    ->status_is(200);
+
+But adding the C<Authorization> header for every request is tedious and likely
+to result in copy/paste errors given enough tests. We'd like to extend
+C<post_ok>'s behavior to build the header for us. With Mojolicious 7.41 (or
+later) and L<Role::Tiny> installed, we can compose this behavior with roles.
+
+First, we create a new package that will implement the role we want:
+
+  package Test::Mojo::Role::BasicAuth;
+  use Role::Tiny;
+  use Mojo::Util 'b64_encode';
+  use Mojo::Base 'Test::Mojo';
+
+  sub post_with_auth_ok {
+      my $t = shift;
+      my $url = shift;
+      my $creds = shift; # username:password
+      my $headers = (ref $_[0] eq 'HASH' ? shift : {});
+      $headers->{Authorization} = "Basic " . b64_encode $creds, '';
+      $t->SUPER::post_ok($url, $headers, @_);
+  }
+
+  1;
+
+This role accepts an extra credentials parameter, then uses that to create an
+C<Authorization> header, then we pass everything along to C<post_ok> to do the
+heavy lifting.
+
+In our test file, we apply the role using C<with_roles> (available in
+Mojolicious 7:41 and later), which creates a new class whose objects have the
+C<post_with_auth_ok> capability:
+
+  my $t = Test::Mojo->with_roles('Test::Mojo::Role::BasicAuth')->new('MyApp');
+
+  $t->post_with_auth_ok('/thermostat/heat', 'me:secretly',
```

Comment:
Yeah, I agree. When I realized that you don't get attributes with Role::Tiny, it makes this no better than the built-in, since you have to supply the credentials with every test. I'll pare this down in the next commit.